### PR TITLE
Extract Runtime Constants into Separate Crates

### PR DESCRIPTION
**Pull request by Doordashcon**
Extracted runtime constants into separate crates, as stated in #4251

cumulus companion: [paritytech/cumulus#849](https://github.com/paritytech/cumulus/pull/849) 

Polkadot address:  12zsKEDVcHpKEWb99iFt3xrTCQQXZMu477nJQsTBBrof5k2h

### CODE REVIEW

File: kusama-runtime-constants/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+[package]
+name = "kusama-runtime-constants"
+version = "0.1.0"
+edition = "2018"
```
Comment (shawntabrizi) [inline kusama-runtime-constants/Cargo.toml:4]: look at the other cargo.toml files to see the metadata we expect

File: Cargo.toml [hunk 2]
```diff
@@ -27,6 +27,11 @@ members = [
 	"core-primitives",
 	"erasure-coding",
 	"primitives",
+	"polkadot-runtime-constants",
+	"kusama-runtime-constants",
+	"rococo-runtime-constants",
+	"test-runtime-constants",
+	"westend-runtime-constants",
```
Comment (bkchr) [inline Cargo.toml:34]: They should go into the runtime folder under `runtime/westend-constants` etc.

File: westend-runtime-constants/src/lib.rs [hunk 3]
```diff
@@ -14,6 +14,7 @@
 // You should have received a copy of the GNU General Public License
 // along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
 
+#![cfg_attr(not(feature = "std"), no_std)]
 /// Money matters.
```
Comment (bkchr) [inline westend-runtime-constants/src/lib.rs:18]:
```suggestion
/// Money matters.
#![cfg_attr(not(feature = "std"), no_std)]
```

**Review by bkchr — CHANGES_REQUESTED** [review]

File: kusama-runtime-constants/Cargo.toml (hunk 1 above)
Comment (Doordashcon) [reply kusama-runtime-constants/Cargo.toml]: fixed

File: Cargo.toml (hunk 2 above)
Comment (Doordashcon) [reply Cargo.toml:34]: They are already in their respective runtime folders as dependencies, how do you mean?

File: westend-runtime-constants/src/lib.rs (hunk 3 above)
Comment (Doordashcon) [reply westend-runtime-constants/src/lib.rs:18]: `cargo +nightly check` returns

```rust
error: an inner attribute is not permitted following an outer doc comment
  --> rococo-runtime-constants/src/lib.rs:18:1
   |
17 |   /// Money matters.
   |   ------------------ previous doc comment
18 |   #![cfg_attr(not(feature = "std"), no_std)]
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not permitted following an outer attribute
```

File: kusama-runtime-constants/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "kusama-runtime-constants"
+version = "0.9.13"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2018"
```
[the remark this answers is not in the document]
Comment (Doordashcon) [reply kusama-runtime-constants/Cargo.toml]: this has been updated

File: Cargo.toml (hunk 2 above)
Comment (apopiak) [reply Cargo.toml:34]: He means the location in the directory structure. You put them as top-level folders, when they should be subdirectories of the runtime crates that depend on them. (so the location of the `Cargo.toml` should IMO `runtime/westend/constants/Cargo.toml`)

Comment (Doordashcon) [reply Cargo.toml]: @apopiak & @bkchr does that solve it?

File: runtime/test-runtime/constants/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "test-runtime-constants"
+version = "0.9.13"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (bkchr) [inline runtime/test-runtime/constants/Cargo.toml:7]:
```suggestion
```

File: runtime/rococo/constants/src/lib.rs [hunk 6]
```diff
@@ -14,6 +14,7 @@
 // You should have received a copy of the GNU General Public License
 // along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
 
+#![cfg_attr(not(feature = "std"), no_std)]
```
Comment (bkchr) [inline runtime/rococo/constants/src/lib.rs:17]:
```suggestion
#![cfg_attr(not(feature = "std"), no_std)]

```

File: runtime/rococo/constants/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "rococo-runtime-constants"
+version = "0.9.13"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (bkchr) [inline runtime/rococo/constants/Cargo.toml:7]:
```suggestion
```

File: runtime/polkadot/constants/src/lib.rs [hunk 8]
```diff
@@ -14,6 +14,7 @@
 // You should have received a copy of the GNU General Public License
 // along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
 
+#![cfg_attr(not(feature = "std"), no_std)]
```
Comment (bkchr) [inline runtime/polkadot/constants/src/lib.rs:17]:
```suggestion
#![cfg_attr(not(feature = "std"), no_std)]

```

File: runtime/polkadot/constants/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "polkadot-runtime-constants"
+version = "0.9.13"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (bkchr) [inline runtime/polkadot/constants/Cargo.toml:7]:
```suggestion
```

File: runtime/kusama/constants/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "kusama-runtime-constants"
+version = "0.9.13"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (bkchr) [inline runtime/kusama/constants/Cargo.toml:7]:
```suggestion
```

File: node/service/Cargo.toml [hunk 11]
```diff
@@ -84,6 +84,12 @@ polkadot-node-subsystem-util = { path = "../subsystem-util" }
 polkadot-runtime-parachains = { path = "../../runtime/parachains" }
 polkadot-node-network-protocol = { path = "../network/protocol" }
 
+# Polkadot Runtime Constants
+polkadot-runtime-constants = { path = "../../runtime/polkadot/constants" }
+kusama-runtime-constants = { path = "../../runtime/kusama/constants" }
+rococo-runtime-constants = { path = "../../runtime/rococo/constants" }
+westend-runtime-constants = { path = "../../runtime/westend/constants" }
```
[the remark this answers is not in the document]
Comment (Doordashcon) [reply node/service/Cargo.toml:91]: the constant value `UNITS` is required from all cause they exist apart from their respective runtimes. would i need to remove the `#[cfg(...)]` above each constant import?

[the remark this answers is not in the document]
Comment (bkchr) [reply node/service/Cargo.toml:91]: No it is about making these deps here optional (`optional = true`) and to only activate them if the feature is activated for the runtime. Just check the runtime crates below.

[the remark this answers is not in the document]
Comment (Doordashcon) [reply node/service/Cargo.toml]: Done

**Comment by Doordashcon** [comment]
@bkchr @apopiak would i need to create a companion PR for cumulus, to address the failing test?

**Comment by shawntabrizi** [comment]
@Doordashcon, I think to resolve this dependency error, you need to re-export the constants out of the expected crate, or just make a companion PR to fix the issue.

See here: https://github.com/paritytech/substrate/blob/master/docs/CONTRIBUTING.adoc#updating-polkadot-as-well

File: runtime/kusama/Cargo.toml [hunk 12]
```diff
@@ -19,6 +19,7 @@ smallvec = "1.6.1"
 authority-discovery-primitives = { package = "sp-authority-discovery", git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
 babe-primitives = { package = "sp-consensus-babe", git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
 beefy-primitives = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
+kusama-runtime-constants = { package = "kusama-runtime-constants", path = "./constants", default-features = false }
```
Comment (bkchr) [inline runtime/kusama/Cargo.toml:22]: You need to enable the std feature below in the file.

File: runtime/test-runtime/Cargo.toml [hunk 13]
```diff
@@ -50,6 +50,8 @@ pallet-staking = { git = "https://github.com/paritytech/substrate", branch = "ma
 pallet-staking-reward-curve = { git = "https://github.com/paritytech/substrate", branch = "master" }
 frame-system = {git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
 frame-system-rpc-runtime-api = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
+test-runtime-constants = { package = "test-runtime-constants", path = "./constants", default-features = false }
```
Comment (bkchr) [inline runtime/test-runtime/Cargo.toml:53]: std feature.

File: runtime/polkadot/Cargo.toml [hunk 14]
```diff
@@ -63,6 +63,7 @@ pallet-staking = { git = "https://github.com/paritytech/substrate", branch = "ma
 pallet-staking-reward-curve = { git = "https://github.com/paritytech/substrate", branch = "master" }
 frame-system = {git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
 frame-system-rpc-runtime-api = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
+polkadot-runtime-constants = { package = "polkadot-runtime-constants", path = "./constants", default-features = false }
```
Comment (bkchr) [inline runtime/polkadot/Cargo.toml:66]: std feature.

File: runtime/westend/Cargo.toml [hunk 15]
```diff
@@ -37,6 +37,8 @@ frame-executive = { git = "https://github.com/paritytech/substrate", branch = "m
 frame-support = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
 frame-system = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
 frame-system-rpc-runtime-api = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
+westend-runtime-constants = { package = "westend-runtime-constants", path = "./constants", default-features = false }
```
Comment (bkchr) [inline runtime/westend/Cargo.toml:40]: std feature

File: runtime/westend/Cargo.toml [hunk 16]
```diff
@@ -37,6 +37,8 @@ frame-executive = { git = "https://github.com/paritytech/substrate", branch = "m
...
+
```
Comment (bkchr) [inline runtime/westend/Cargo.toml:41]:
```suggestion
```

File: runtime/test-runtime/Cargo.toml [hunk 17]
```diff
@@ -50,6 +50,8 @@ pallet-staking = { git = "https://github.com/paritytech/substrate", branch = "ma
...
+
```
Comment (bkchr) [inline runtime/test-runtime/Cargo.toml:54]:
```suggestion
```

**Review by bkchr — APPROVED** [review]
Ty! Companion also needs to be done.

File: runtime/kusama/Cargo.toml (hunk 12 above)
Comment (Doordashcon) [reply runtime/kusama/Cargo.toml:22]: Done

File: runtime/polkadot/Cargo.toml (hunk 14 above)
Comment (Doordashcon) [reply runtime/polkadot/Cargo.toml:66]: Done

File: runtime/westend/Cargo.toml (hunk 15 above)
Comment (Doordashcon) [reply runtime/westend/Cargo.toml:40]: Done

File: runtime/test-runtime/Cargo.toml (hunk 13 above)
Comment (Doordashcon) [reply runtime/test-runtime/Cargo.toml:53]: Done

**Review by shawntabrizi — APPROVED** [review]

**Comment by Doordashcon** [comment]
Having some trouble with cumulus companion

**Comment by Doordashcon** [comment]
Got an email notification about a tip? @shawntabrizi

**Comment by Doordashcon** [comment]
There was an extra space after the colon, fix

**Comment by shawntabrizi** [comment]
Thanks for your patience @Doordashcon

**Comment by Doordashcon** [comment]
> Thanks for your patience @Doordashcon

Thanks for the tip

**Pull request merged**